src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,182 @@
+        if (!JWTSigningKey.isJWTAlgorithmSupported(algorithm)) {
+            throw new NoSuchAlgorithmException(algorithm + "is not a supported signing algorithm");
+        }
+
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+
+        if (storage.getType() == STORAGE_TYPE.SQL) {
+            SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
+
+            return sqlStorage.startTransaction(con -> {
+               JWTSigningKeyInfo keyInfo;
+               String algorithmType;
+
+               try {
+                   algorithmType = getAlgorithmType(algorithm);
+               } catch (NoSuchAlgorithmException e) {
+                   throw new StorageTransactionLogicException(e);
+               }
+
+               JWTSigningKeyInfo keyFromStorage = sqlStorage.getLatestJWTSigningKeyForAlgorithm_Transaction(con, algorithmType);
+
+               if (keyFromStorage == null) {
+                   try {
+                       long currentTimeInMillis = System.currentTimeMillis();
+                       Utils.PubPriKey newKey = Utils.generateNewPubPriKey();
+                       keyInfo = new JWTSigningKeyInfo(Utils.getUUID(), newKey.publicKey, newKey.privateKey, currentTimeInMillis, algorithm, algorithmType);
+                       sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                   } catch (NoSuchAlgorithmException e) {
+                       throw new StorageTransactionLogicException(e);
+                   }
+               } else {
+                   keyInfo = keyFromStorage;
+               }
+
+               sqlStorage.commitTransaction(con);
+               return keyInfo;
+            });
+        } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
+            SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
+
+            JWTSigningKeyInfo keyInfo;
+
+            String algorithmType;
+
+            try {
+                algorithmType = getAlgorithmType(algorithm);
+            } catch (NoSuchAlgorithmException e) {
+                throw new StorageTransactionLogicException(e);
+            }
+
+            JWTSigningKeyInfo keyFromStorage = noSQLStorage.getLatestJWTSigningKeyForAlgorithm_Transaction(algorithmType);
+
+            if (keyFromStorage == null) {
+                long currentTimeInMillis = System.currentTimeMillis();
+                Utils.PubPriKey newKey = Utils.generateNewPubPriKey();
+                keyInfo = new JWTSigningKeyInfo(Utils.getUUID(), newKey.publicKey, newKey.privateKey, currentTimeInMillis, algorithm, algorithmType);
+            } else {
+                keyInfo = keyFromStorage;
+            }
+
+            return keyInfo;
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    private String getAlgorithmType(String algorithm) throws NoSuchAlgorithmException {
```

Comment:
Algorithm handling has been moved to an enum `JWTSigningKey.SupportedAlgorithms`, the enum now exposes a way to get algorithm strings and types instead of having functions like this one.